src/Evolution/Systems/RadiationTransport/M1Grey/M1Closure.cpp

```diff
@@ -0,0 +1,270 @@
+
+      double& xi = get(*closure_factor)[s];
+      // Root finding function
+      const auto XiJSqrMinusH2 = [
+        &e_pt, &j_0, &j_thin, &j_thick, &hh_0, &hh_thick, &hh_thin,
+        &hh_thin_thin, &hh_thick_thick, &hh_thin_thick
+      ](double lxi) noexcept {
+        const double chi = M1ClosureFunction(lxi);
+        const double dchi_dxi = M1ClosureDeriv(lxi);
+        const double dthin = 1.5 * chi - 0.5;
+        const double dthick = 1. - dthin;
+        const double dthin_dxi = 1.5 * dchi_dxi;
+        const double dthick_dxi = -dthin_dxi;
+
+        const double e_fluid = j_0 + j_thin * dthin + j_thick * dthick;
+        const double de_fluid_dxi = j_thin * dthin_dxi + j_thick * dthick_dxi;
+        const double h2 = hh_0 + hh_thick * dthick + hh_thin * dthin +
+                          hh_thin_thin * square(dthin) +
+                          hh_thick_thick * square(dthick) +
+                          hh_thin_thick * dthin * dthick;
+        const double dh2_ddthin =
+            hh_thin + hh_thin_thick * dthick + 2. * hh_thin_thin * dthin;
+        const double dh2_ddthick =
+            hh_thick + hh_thin_thick * dthin + 2. * hh_thick_thick * dthick;
+        return std::make_pair(
+            (square(e_fluid * lxi) - h2) / square(e_pt),
+            (2. * e_fluid * de_fluid_dxi * square(lxi) +
+             2. * square(e_fluid) * lxi - dthin_dxi * dh2_ddthin -
+             dthin_dxi * dh2_ddthick) /
+                square(e_pt));
+      };
+      // Initial guess for root finding.
+      // Choice 1: previous value
+      // Choice 2: value for zero-velocity
+      double xi_guess = xi;
+      if (xi <= TINY || xi >= 1.) {
+        xi_guess = sqrt(f2_pt) / e_pt;
+      }
+      // Rootfinding - currently use Newton-Raphson - what accuracy do we need?
+      xi = RootFinder::newton_raphson(XiJSqrMinusH2, xi_guess, 0., 1., 6);
+
+      // Assemble output quantities:
+      const double chi = M1ClosureFunction(xi);
+      const double dthin = 1.5 * chi - 0.5;
+      const double dthick = 1. - dthin;
+      get(*comoving_energy_density)[s] =
+          j_0 + j_thin * dthin + j_thick * dthick;
+      get(*comoving_momentum_density_normal)[s] =
+          h_0_n + h_thin_n * dthin + h_thick_n * dthick;
+      for (int i = 0; i < 3; i++) {
+        comoving_momentum_density_spatial->get(i) =
+            -(h_0_v + h_thin_v * dthin + h_thick_v * dthick) * v_low.get(i)[s] -
+            (h_0_f + h_thin_f * dthin + h_thick_f * dthick) *
+                momentum_density.get(i)[s];
+        for (int j = i; j < 3; j++) {
+          // Optically thin part of pressure tensor
+          pressure_tensor->get(i, j) = dthin * e_pt *
+                                       momentum_density.get(i)[s] *
+                                       momentum_density.get(j)[s] / f2_pt;
+        }
+      }
+      // Optically thick limit
+      const double J = 3. / (2. * w2_pt + 1.) *
```

Comment:
move this to right before line 254, make this J_over_3 to avoid the multiplication and division by 3.